src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs

```diff
@@ -1670,14 +1673,23 @@ internal static BoundBlock BindMethodBody(MethodSymbol method, TypeCompilationSt
                     BoundNode methodBodyForSemanticModel = methodBody;
                     NullableWalker.SnapshotManager snapshotManager = null;
                     ImmutableDictionary<Symbol, Symbol> remappedSymbols = null;
-                    if (bodyBinder.Compilation.NullableSemanticAnalysisEnabled)
+                    var compilation = bodyBinder.Compilation;
+                    var isSufficientLangVersion = compilation.LanguageVersion >= MessageID.IDS_FeatureNullableReferenceTypes.RequiredVersion();
+                    switch (isSufficientLangVersion, compilation)
                     {
-                        // Currently, we're passing an empty DiagnosticBag here because the flow analysis pass later will
-                        // also run the nullable walker, and issue duplicate warnings. We should try to only run the pass
-                        // once.
-                        // https://github.com/dotnet/roslyn/issues/35041
-                        methodBodyForSemanticModel = NullableWalker.AnalyzeAndRewrite(bodyBinder.Compilation, method, methodBody, bodyBinder, new DiagnosticBag(), createSnapshots: true, out snapshotManager, ref remappedSymbols);
+                        case (true, { NullableSemanticAnalysisEnabled: true }):
+                            methodBodyForSemanticModel = NullableWalker.AnalyzeAndRewrite(compilation, method, methodBody, bodyBinder, diagnostics, createSnapshots: true, out snapshotManager, ref remappedSymbols);
+                            break;
+                        case (true, { ShouldRunNullableWalker: true }):
```

Comment:
>ShouldRunNullableWalker [](start = 38, length = 23)

Is this checked in ```NullableWalker.AnalyzeIfNeeded(_compilation, methodSymbol, analyzedInitializers, diagsForCurrentMethod);```? If so, consider restructuring the code to use this helper here as well and avoiding duplicate checks.